Null check scriptExecutionContext before using it in DOMCacheStorage.cpp

cdumez opened the pull request:
#### 656c046fbbef93f936aa87d142d3ba8e8403cff1
<pre>
Null check scriptExecutionContext before using it in DOMCacheStorage.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=268625">https://bugs.webkit.org/show_bug.cgi?id=268625</a>
<a href="https://rdar.apple.com/122170377">rdar://122170377</a>

Reviewed by Sihui Liu, Brent Fulgham and Youenn Fablet.

* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::create):
(WebCore::DOMCache::DOMCache):
* Source/WebCore/Modules/cache/DOMCache.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
(WebCore::DOMCacheStorage::findCacheOrCreate):
(WebCore::DOMCacheStorage::doOpen):

Canonical link: <a href="https://commits.webkit.org/274013@main">https://commits.webkit.org/274013@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ea416568e78493d1b92b4a4bb866b56892918cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37513 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16397 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39776 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40039 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33430 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38827 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19016 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13561 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31825 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38078 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13827 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/32908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12041 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12014 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33592 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41302 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33912 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37910 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12579 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10142 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36071 "Found 1 new API test failure: /WebKitGTK/TestLoaderClient:/webkit/WebKitWebView/load-twice-and-reload (failure)") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/14034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8456 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12985 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/Modules/cache/DOMCache.cpp:
```diff
@@ -41,15 +41,15 @@
-DOMCache::DOMCache(ScriptExecutionContext& context, String&& name, DOMCacheIdentifier identifier, Ref<CacheStorageConnection>&& connection)
-    : ActiveDOMObject(&context)
+DOMCache::DOMCache(ScriptExecutionContext* context, String&& name, DOMCacheIdentifier identifier, Ref<CacheStorageConnection>&& connection)
+    : ActiveDOMObject(context)
```
It's surprising that `DOMCache` can exist in the absence of an active `ScriptExecutionContext`, but I see that `ActiveDOMObject` does expect this, so this seems correct.

cdumez replied:
Yes, having a null scriptExecution context is something that is expected/supported by ActiveDOMObject.

brentfulgham reviewed (approved):
LGTM

szewai commented on Source/WebCore/Modules/cache/DOMCache.cpp:
```diff
@@ -41,15 +41,15 @@
 using namespace WebCore::DOMCacheEngine;
 
-Ref<DOMCache> DOMCache::create(ScriptExecutionContext& context, String&& name, DOMCacheIdentifier identifier, Ref<CacheStorageConnection>&& connection)
+Ref<DOMCache> DOMCache::create(ScriptExecutionContext* context, String&& name, DOMCacheIdentifier identifier, Ref<CacheStorageConnection>&& connection)
```
I think this should still be a `ScriptExecutionContext&`, because if the script context is null then the object should not be created. We should probably add a check for `ScriptExecutionContext` in `DOMCacheStorage` and throw exception if the context is null.

cdumez replied:
Ok, we can also do that. I hesitated between the 2.

cdumez replied:
I'll just point out that your approach is more painful (requires a bit of refactoring at call sites) and I don't think there are major benefits, it just "feels" cleaner. See `findCacheOrCreate()` for example, currently returns a `Ref<DOMCache>` so it is never expected to fail.

szewai replied:
Thanks, adding a null check seems the right thing to do (given that there is no guarantee context() will not return null in previous code)

cdumez committed:
Null check scriptExecutionContext before using it in DOMCacheStorage.cpp
https://bugs.webkit.org/show_bug.cgi?id=268625
rdar://122170377

Reviewed by Sihui Liu, Brent Fulgham and Youenn Fablet.

* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::create):
(WebCore::DOMCache::DOMCache):
* Source/WebCore/Modules/cache/DOMCache.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
(WebCore::DOMCacheStorage::findCacheOrCreate):
(WebCore::DOMCacheStorage::doOpen):

Canonical link: https://commits.webkit.org/274013@main
